more_itertools/recipes.py

```diff
@@ -346,6 +347,116 @@ def powerset(iterable):
     return chain.from_iterable(combinations(s, r) for r in range(len(s) + 1))
 
 
+def unique_powerset(iterable):
+    """
+    The unique_powerset(iterable) returns the unique combinations of values
+    when presented with repeated values:
+
+        unique_powerset([1,1,1,2,2,3]) --> [
+            (), (1,), (1, 1), (1, 1, 1), (2,), (2, 2), (3,),
+            (1, 2), (1, 1, 2), (1, 1, 1, 2),
+            (1, 2, 2), (1, 1, 2, 2), (1, 1, 1, 2, 2),
+            (1, 3), (1, 1, 3), (1, 1, 1, 3),
+            (2, 3), (2, 2, 3), (1, 2, 3),
+            (1, 1, 2, 3), (1, 1, 1, 2, 3),
+            (1, 2, 2, 3), (1, 1, 2, 2, 3),
+            (1, 1, 1, 2, 2, 3)
+        ] # 23 records
+
+    This should be viewed in contrast to the powerset (see [1]) which would
+    generate repeated values:
+
+        powerset([1,1,1,2,2,3]) --> [
+            (),
+            (1,), (1,), (1,),
+            (2,), (2,),
+            (3,),
+            (1, 1), (1, 1), (1, 1),
+            (1, 2), (1, 2), (1, 2), (1, 2), (1, 2), (1, 2),
+            (1, 3), (1, 3), (1, 3),
+            ... cut for brevity ...
+            (1, 1, 1, 2, 2), (1, 1, 1, 2, 3), (1, 1, 1, 2, 3),
+            (1, 1, 2, 2, 3), (1, 1, 2, 2, 3), (1, 1, 2, 2, 3)
+        ] # 63 records.
+
+    The application of the unique_powerset is plenty, but most dominantly in
+    the knapsack problem where a number of items must be matched to the
+    capacity limit of the knapsack. Each combination of items determine the
+    utilisation of the knapsack and the best should be selected, yet items
+    of the same value can be substituted without further consideration
+
+    As the number of duplicate values grow, the number of redundant options
+    grows exponentially if using powerset.
+    In the example above the powerset generates 63 vs the 24 unique options
+    generated in the unique_powerset.
+
+    The assertion set(powerset(iterable)) == unique_powerset(iterable) must
+    always be true, and whilst the former method is available, powerset of
+    any iterables longer than 20 items, become intolerable except for the most
+    patient programmers.
+
+    [1] https://docs.python.org/3/library/itertools.html#itertools-recipes
+
+    :param iterable:
+    :return: list of tuples
+    """
+    yield ()
+    # first we summarize the iterable into blocks of identical values. Example:
+    # [1,1,1,2,2,3] -->
+    # d = {
+    #     1: [[1],[1,1],[1,1,1]],
+    #     2: [[2],[2,2]],
+    #     3: [[3]]
+    #     }
+    d = {i: iterable.count(i) for i in set(iterable)}
+    blocks = {i: [] for i in set(iterable)}
```

Comment:
Here's an issue I noticed: this exhausts `iterable` once on line 411. If it's an iterator and not a sequence, then the second time `set(iterable)` is attempted it will fail.

Reply:
We can outsource the creation of these dicts to the `collections` module:

```python
    blocks = defaultdict(list)
    for k, v in Counter(iterable).items():
        for i in range(1, v + 1):
            blocks[k].append([k] * i)
```